### PR TITLE
Add groundwater transport capabilities to nlmod

### PR DESCRIPTION
**Pull request by dbrakenhoff**
This PR introduces functionality to create a transport model using `nlmod`.

- Add BUY package to GWF
- Add gwt module for building groundwater transport model
- Add gwt.output for post-processing concentrations (and converting pointwater and freshwater heads)
- Add gwt.prepare module to set dataset attributes with default values for transport modeling (currently only for chloride)
- Improve GHB and CHD constructors to accept auxiliary variables
- Improve `lrc_to_reclist` and `lcid_to_reclist` to accept aux vars
- Add `aggregate_by_weighted_mean_to_ds` to vertically calculate the weighted mean per layer from a dataset with different layer definition.
- Add example notebook.

### CODE REVIEW

File: nlmod/gwt/gwt.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+import flopy
+import xarray as xr
+
+from ..dims import grid
+from ..gwf.gwf import _dis, _disv, _set_record
+
+logger = logging.getLogger(__name__)
+
+
+def _get_var_from_ds_attr(ds, varname, attr=None, var=None, warn=True):
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:13]: It took me a while to understand what this function does. I think a better name would be `_get_value_from_ds_attr`, curious what you think about that?

Is the idea that this is a general function that will be used in the future in gwf?

File: nlmod/gwt/gwt.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        dataset containing model data
+    varname : str
+        name of the variable
+    attr : str, optional
+        name of the attribute in dataset (is sometimes different to varname)
+    var : Any, optional
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:24]: var -> value

File: nlmod/gwt/gwt.py [hunk 3]
```diff
@@ -0,0 +1,422 @@
...
+            msg = (
+                f"No value found for '{varname}', passing None to flopy. "
+                f"To fix this error pass '{varname}' to function or set 'ds.{attr}'."
+            )
+            logger.warning(msg)
+        # raise ValueError(msg)
+    return var
+
+
+def _get_var_from_ds_datavar(ds, varname, datavar=None, var=None, warn=True):
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:55]: `_get_value_from_ds_datavar`?

File: nlmod/gwt/gwt.py [hunk 4]
```diff
@@ -0,0 +1,422 @@
...
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        dataset containing model data
+    varname : str
+        name of the variable
+    datavar : str, optional
+        name of the data variable (is sometimes different to varname) in dataset
+    var : Any, optional
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:66]: var -> value

File: nlmod/gwt/gwt.py [hunk 5]
```diff
@@ -0,0 +1,422 @@
...
+        variable value, by default None
+    warn : bool, optional
+        log warning if value not found
+
+    Returns
+    -------
+    var : Any
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:73]: var -> value

File: nlmod/gwt/gwt.py [hunk 6]
```diff
@@ -0,0 +1,422 @@
...
+
+def _get_var_from_ds_attr(ds, varname, attr=None, var=None, warn=True):
+    """Internal function to parse variable from dataset attributes.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        dataset containing model data
+    varname : str
+        name of the variable
```
Comment (OnnoEbbens) [inline nlmod/gwt/gwt.py:21]: name of the variable in a flopy package?

File: nlmod/gwt/output.py [hunk 7]
```diff
@@ -0,0 +1,257 @@
...
+
+    Note: Calling this function with ds is currently preferred over calling it
+    with gwf, because the layer and time coordinates can not be fully
+    reconstructed from gwf.
+
+    Parameters
+    ----------
+    ds : xarray.Dataset
+        Xarray dataset with model data.
+    gwt : flopy ModflowGwf
```
Comment (OnnoEbbens) [inline nlmod/gwt/output.py:40]: flopy ModflowGwf should flopy ModflowGwt I think?

File: nlmod/gwt/output.py [hunk 8]
```diff
@@ -0,0 +1,257 @@
...
+    if fname_conc is not None:
+        concobj = flopy.utils.HeadFile(fname_conc, text="concentration")
+    return concobj
+
+
+def get_concentration_da(ds=None, gwt=None, fname_conc=None):
+    """Reads concentration file given either a dataset or a groundwater flow object.
+
+    Note: Calling this function with ds is currently preferred over calling it
+    with gwf, because the layer and time coordinates can not be fully
```
Comment (OnnoEbbens) [inline nlmod/gwt/output.py:33]: with gwf -> with gwt

File: nlmod/gwt/prepare.py [hunk 9]
```diff
@@ -0,0 +1,60 @@
+def set_default_transport_parameters(ds, transport_type):
```
Comment (OnnoEbbens) [inline nlmod/gwt/prepare.py:1]: Can also add this function to gwt.py and not make a separate module?

File: nlmod/version.py [hunk 10]
```diff
@@ -1 +1,19 @@
...
+
+def show_versions() -> None:
+    """Method to print the version of dependencies."""
+
+    msg = (
+        f"Python version: {python_version()}\n"
+        f"NumPy version: {metadata.version('numpy')}\n"
+        f"Xarray version: {metadata.version('xarray')}\n"
+        f"Matplotlib version: {metadata.version('matplotlib')}\n"
+    )
```
Comment (OnnoEbbens) [inline nlmod/version.py:15]: add flopy?

**Review by OnnoEbbens — APPROVED** [review]
Good job! I added some questions/ideas

File: nlmod/gwt/gwt.py (hunk 1 above)
Comment (dbrakenhoff) [reply nlmod/gwt/gwt.py:13]: Good suggestion, I agree about the name change. And yes, this was a first attempt at a more general approach for getting data from the model dataset in our model building functions.

File: nlmod/gwt/gwt.py (hunk 6 above)
Comment (dbrakenhoff) [reply nlmod/gwt/gwt.py:21]: :+1:

File: nlmod/gwt/gwt.py (hunk 3 above)
Comment (dbrakenhoff) [reply nlmod/gwt/gwt.py:55]: :+1:

File: nlmod/gwt/gwt.py (hunk 5 above)
Comment (dbrakenhoff) [reply nlmod/gwt/gwt.py:73]: done.

File: nlmod/gwt/gwt.py (hunk 4 above)
Comment (dbrakenhoff) [reply nlmod/gwt/gwt.py:66]: done.

File: docs/examples/02_surface_water.ipynb [hunk 11]
```diff
@@ -797,8 +797,27 @@
   }
  ],
  "metadata": {
+  "kernelspec": {
```
[the remark this answers is not in the document]
Comment (dbrakenhoff) [reply docs/examples/02_surface_water.ipynb]: Fixed

File: nlmod/gwt/prepare.py (hunk 9 above)
Comment (dbrakenhoff) [reply nlmod/gwt/prepare.py:1]: I like it being somewhat separate even if it is a very small module.

File: nlmod/version.py (hunk 10 above)
Comment (dbrakenhoff) [reply nlmod/version.py:15]: Good idea

File: nlmod/gwf/gwf.py [hunk 12]
```diff
@@ -56,21 +56,43 @@ def dis(ds, gwf, length_units="METERS", pname="dis", **kwargs):
...
     Returns
     -------
-    dis : TYPE
+    dis : flopy ModflowGwfdis
+        discretisation package.
+    """
+    return _dis(ds, gwf, length_units, pname, **kwargs)
+
+
+def _dis(ds, model, length_units="METERS", pname="dis", **kwargs):
```
[the remark this answers is not in the document]
Comment (dbrakenhoff) [reply nlmod/gwf/gwf.py:73]: This way the docstrings are more correct :)

File: nlmod/dims/layers.py [hunk 13]
```diff
@@ -1078,3 +1078,68 @@ def update_idomain_from_thickness(idomain, thickness, mask):
             idomain[ilay] = xr.where(mask3, 1, idomain[ilay])
 
     return idomain
+
+
+def aggregate_by_weighted_mean_to_ds(ds, source_ds, var_name):
```
[the remark this answers is not in the document]
Comment (dbrakenhoff) [reply nlmod/dims/layers.py:1083]: Yea, wasn´t too sure either, but I figured that resample is more for horizontal (xy-plane) resampling and layers is the place for vertical operations...?

Leaving it here for now.

**Pull request merged**